src/target/target_id.cc

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/target/target_id.cc
+ * \brief Target id registry
+ */
+#include <tvm/target/target_id.h>
+
+#include "../node/attr_registry.h"
+#include "../runtime/object_internal.h"
+
+namespace tvm {
+
+using TargetIdRegistry = AttrRegistry<TargetIdRegEntry, TargetId>;
+
+TVM_DLL TargetIdRegEntry& TargetIdRegEntry::RegisterOrGet(const String& target_id_name) {
+  return TargetIdRegistry::Global()->RegisterOrGet(target_id_name);
+}
+
+TVM_DLL void TargetIdRegEntry::UpdateAttr(const String& key, TVMRetValue value, int plevel) {
+  TargetIdRegistry::Global()->UpdateAttr(key, id_, value, plevel);
+}
+
+TVM_DLL const AttrRegistryMapContainerMap<TargetId>& TargetId::GetAttrMapContainer(
+    const String& attr_name) {
+  return TargetIdRegistry::Global()->GetAttrMap(attr_name);
+}
+
+TVM_DLL const TargetId& TargetId::Get(const String& target_id_name) {
+  const TargetIdRegEntry* reg = TargetIdRegistry::Global()->Get(target_id_name);
+  CHECK(reg != nullptr) << "TargetId " << target_id_name << " is not registered";
+  return reg->id_;
+}
+
+void VerifyTypeInfo(const ObjectRef& obj, const TargetIdNode::ValueTypeInfo& info) {
+  CHECK(obj.defined()) << "Object is None";
```

Comment:
We will need to add a second level validation (e.g. in the case of target host, the result is expected to be target. So the map is expected to follow the schema of the corresponding target. aka. Target is a special type that is more fine grained than the object type system.